<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rPHG">
<title>Getting Started with rPHG • rPHG</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with rPHG">
<meta property="og:description" content="rPHG">
<meta property="og:image" content="https://maize-genetics.github.io/rPHG/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rPHG</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.13</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/rphg_walkthrough.html">Getting Started with rPHG</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Getting Started with rPHG</h1>
                        <h4 data-toc-skip class="author">Brandon
Monier</h4>
            
            <h4 data-toc-skip class="date">2019-12-05 at 13:21:26</h4>
      
      
      <div class="d-none name"><code>rphg_walkthrough.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>Thanks for checking out rPHG! In this document, we will go over the
functionalities used to work with the practical haplotype graph (PHG)
API via R.</p>
<p>The PHG is a trellis graph based representation of genic and
intergenic regions (called reference ranges or reference intervals)
which represent diversity across and between taxa. It can be used to:
create custom genomes for alignment, call rare alleles, impute
genotypes, and efficiently store genomic data from many lines
(i.e. reference, assemblies, and other lines). Skim sequences generated
for a given taxon are aligned to consensus sequences in the PHG to
identify the haplotype node at a given anchor. All the anchors for a
given taxon are processed through a Hidden Markov Model (HMM) to
identify the most likely path through the graph. Path information is
used to identify the variants (SNPs). Low cost sequencing technologies,
coupled with the PHG, facilitate the genotyping of large number of
samples to increase the size of training populations for genomic
selection models. This can in turn increase predictive accuracy and
selection intensity in a breeding program.</p>
<p>Detailed documentation and source code can be found on our
website:</p>
<p><a href="https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/Home" class="external-link uri">https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/Home</a></p>
</div>
<div class="section level3">
<h3 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h3>
<p>The main goal of developing this package is to construct an R-based
front-end to build and interact with the PHG API that implements
commonly used Biocondcutor classes, data structures, and accessor
methods for downstream analysis and integration with other packages.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation-and-preliminary-steps">Installation and Preliminary Steps<a class="anchor" aria-label="anchor" href="#installation-and-preliminary-steps"></a>
</h2>
<div class="section level3">
<h3 id="prerequisites---installing-rjava">Prerequisites - installing rJava<a class="anchor" aria-label="anchor" href="#prerequisites---installing-rjava"></a>
</h3>
<p>Since the PHG is written in Java, Java JDK will need to be installed
on your machine. Additionally, for R to communicate with Java, the R
package <code>rJava</code> will need to be installed. Detailed
information can be found using the following links, depending on your
OS:</p>
<ul>
<li><a href="https://datawookie.netlify.com/blog/2018/02/installing-rjava-on-ubuntu/" class="external-link">Linux</a></li>
<li><a href="https://zhiyzuo.github.io/installation-rJava/" class="external-link">macOS</a></li>
<li><a href="https://cimentadaj.github.io/blog/2018-05-25-installing-rjava-on-windows-10/installing-rjava-on-windows-10/" class="external-link">Windows</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="problems-with-rjava-if-you-have-upgraded-java">Problems with rJava if you have upgraded Java<a class="anchor" aria-label="anchor" href="#problems-with-rjava-if-you-have-upgraded-java"></a>
</h3>
<p>When using macOS, if you previously had <code>rJava</code> working
through RStudio, then you upgraded your Java and it now longer works,
try the following:</p>
<p>At the command line type:</p>
<pre><code>R CMD javareconf</code></pre>
<p>Then check for a left over symbolic link via:</p>
<pre><code><span><span class="va">ls</span> <span class="op">-</span><span class="va">ltr</span> <span class="op">/</span><span class="va">usr</span><span class="op">/</span><span class="va">local</span><span class="op">/</span><span class="va">lib</span><span class="op">/</span><span class="va">libjvm.dylib</span></span></code></pre>
<p>If the link exists, remove it, then create it fresh via these
commands:</p>
<pre><code>rm /usr/local/lib/libjvm.dylib
sudo ln -s $(/usr/libexec/java_home)/lib/server/libjvm.dylib /usr/local/lib</code></pre>
<p>You should now be able to enter RStudio and setup rJava.</p>
</div>
<div class="section level3">
<h3 id="install-from-bitbucket">Install from BitBucket<a class="anchor" aria-label="anchor" href="#install-from-bitbucket"></a>
</h3>
<p>After you have <code>rJava</code> up and running on your machine,
install <code>rPHG</code> by installing the source code from our
BitBucket repository:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_bitbucket</span><span class="op">(</span></span>
<span>    repo <span class="op">=</span> <span class="st">"bucklerlab/rphg"</span>,</span>
<span>    ref <span class="op">=</span> <span class="st">"master"</span>,</span>
<span>    build_vignettes <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>After source code has been compiled, the package can be loaded
using:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bitbucket.org/bucklerlab/rphg/src/master/" class="external-link">rPHG</a></span><span class="op">)</span></span></code></pre></div>
<p>Or, if you want to use a function without violating your environment
you can use <code>rPHG::&lt;function&gt;</code>, where
<code>&lt;function&gt;</code> is an <code>rPHG</code> function.</p>
</div>
<div class="section level3">
<h3 id="preliminary-steps">Preliminary steps<a class="anchor" aria-label="anchor" href="#preliminary-steps"></a>
</h3>
<div class="section level4">
<h4 id="setting-memory">Setting Memory<a class="anchor" aria-label="anchor" href="#setting-memory"></a>
</h4>
<p>Since genome-wide association analyses can use up a lot of
computational resources, memory allocation to <code>rPHG</code> can be
modified. To change the amount of memory, use the base
<code><a href="https://rdrr.io/r/base/options.html" class="external-link">options()</a></code> function and modify the following parameter:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>java.parameters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-Xmx&lt;memory&gt;"</span>, <span class="st">"-Xms&lt;memory&gt;"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Replace <code>&lt;memory&gt;</code> with a specified unit of memory.
For example, if I want to allocate a maximum of 6 GB of memory for my
operations, I would use the input <code>"-Xmx6g"</code>, where
<code>g</code> stands for gigabyte (GB). More information about memory
allocation can be found <a href="https://stackoverflow.com/questions/14763079/what-are-the-xms-and-xmx-parameters-when-starting-jvm" class="external-link">here</a>.</p>
<p><strong>NOTE:</strong> Setting Java memory options for
<code>rPHG</code> and any <code>rJava</code>-related packages
<em>needs</em> to be set <em>before</em> loading the <code>rPHG</code>
package!</p>
</div>
<div class="section level4">
<h4 id="the-importance-of-logging-your-progress">The importance of logging your progress<a class="anchor" aria-label="anchor" href="#the-importance-of-logging-your-progress"></a>
</h4>
<p>Before we begin analyzing data, optional parameters can be set up to
make <code>rPHG</code> more efficient. To prevent your R console from
being overloaded with PHG logging information, <strong>it is highly
recommended that you start a logging file</strong>. This file will house
all of PHG’s logging output which is beneficial for debugging and
tracking the progress of your analytical workflow. To start a logging
file, use the following command:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rPHG</span><span class="fu">::</span><span class="fu"><a href="../reference/startLogger.html">startLogger</a></span><span class="op">(</span>fullPath <span class="op">=</span> <span class="cn">NULL</span>, fileName <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p>If the <code><a href="../reference/startLogger.html">rPHG::startLogger()</a></code> parameters are set to
<code>NULL</code>, the logging file will be created in your current
working directory. If you are unsure of what your working directory is
in R, use the base <code><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd()</a></code> command.</p>
<p>Additionally, since this is a general walkthrough, certain
intricacies of each function may glossed over. If you would like to
study a function in full, refer to the R documentation by using
<code>?&lt;function&gt;</code> in the console, where
<code>&lt;function&gt;</code> is an <code>rPHG</code>-based
function.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="databases-and-configuration-files">Databases and Configuration Files<a class="anchor" aria-label="anchor" href="#databases-and-configuration-files"></a>
</h2>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>The idea behind the PHG is that in a given breeding program, all
parental genotypes can be sequenced at high coverage, and loaded as
parental haplotypes in a relational database. Progeny can then be
sequenced at low coverage and used to infer which parental
haplotypes/genotypes from the database are most likely present in a
given progeny.</p>
</div>
<div class="section level3">
<h3 id="database-types">Database types<a class="anchor" aria-label="anchor" href="#database-types"></a>
</h3>
<p>Currently, the PHG can use SQLite or PostgreSQL to store data for the
pan-genomic graph. For more information about how data is stored within
the database schema, please refer to our <a href="https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/Home" class="external-link">wiki</a>.</p>
</div>
<div class="section level3">
<h3 id="configuration-files">Configuration files<a class="anchor" aria-label="anchor" href="#configuration-files"></a>
</h3>
<p>Access to the PHG database, regardless of database type, requires a
configuration file. This file contains various metadata needed to access
the PHG db and calculate optimal graph paths:</p>
<ul>
<li>host:port</li>
<li>username</li>
<li>password</li>
<li>database name</li>
<li>database type (sqlite or postgres)</li>
</ul>
<p>An example database configuration can be found below:</p>
<p><strong>SQLite example</strong></p>
<pre><code>host=localHost
user=sqlite
password=sqlite
DB=/tempFileDir/outputDir/phgTestDB_mapq48.db
DBtype=sqlite</code></pre>
<p><strong>PostgreSQL example</strong></p>
<pre><code>host=172.17.0.2:5432
user=postgres
password=phgP0stgr3s
DB=phgdb
DBtype=postgres</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="database-access-and-graph-building">Database Access and Graph Building<a class="anchor" aria-label="anchor" href="#database-access-and-graph-building"></a>
</h2>
<div class="section level3">
<h3 id="method-calling">Method calling<a class="anchor" aria-label="anchor" href="#method-calling"></a>
</h3>
<p>Once you have a PHG database and configuration file, you can proceed
to the following steps. First, you can access all availabe PHG methods
from the database using a path parameter to the database configuration
file:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example path location (not run)</span></span>
<span><span class="va">configPath</span> <span class="op">&lt;-</span> <span class="st">"/home/bm646/Temporary/phg_tests/configSQLite.txt"</span></span>
<span></span>
<span><span class="va">phgMethods</span> <span class="op">&lt;-</span> <span class="fu">rPHG</span><span class="fu">::</span><span class="fu"><a href="../reference/showPHGMethods.html">showPHGMethods</a></span><span class="op">(</span>configFile <span class="op">=</span> <span class="va">configPath</span><span class="op">)</span></span>
<span><span class="va">phgMethods</span></span></code></pre></div>
<pre><code><span><span class="co">## # A tibble: 7 x 5</span></span>
<span><span class="co">##   method_id method_type type_name        method_name      description       </span></span>
<span><span class="co">##       &lt;int&gt;       &lt;int&gt; &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;             </span></span>
<span><span class="co">## 1         1           1 ANCHOR_HAPLOTYP~ B73Ref_method    Test version for ~</span></span>
<span><span class="co">## 2         2           7 REF_RANGE_GROUP  refRegionGroup   Group consists of~</span></span>
<span><span class="co">## 3         3           7 REF_RANGE_GROUP  refInterRegionG~ Group consists of~</span></span>
<span><span class="co">## 4         4           2 ASSEMBLY_HAPLOT~ mummer4          Assembly aligned ~</span></span>
<span><span class="co">## 5         5           1 ANCHOR_HAPLOTYP~ GATK_PIPELINE    GATK_PIPELINE cre~</span></span>
<span><span class="co">## 6         6           1 ANCHOR_HAPLOTYP~ CONSENSUS        CONSENSUS for cre~</span></span>
<span><span class="co">## 7         7           1 ANCHOR_HAPLOTYP~ consensusTest    consensusTest;det~</span></span></code></pre>
<p>The above object will produce a <code>tibble</code>-based data frame
that will contain return method data from the PHG database. Method IDs
and descriptions can be viewed for additional information detailing the
methods available to use when building the PHG graph object.</p>
</div>
<div class="section level3">
<h3 id="graph-building">Graph building<a class="anchor" aria-label="anchor" href="#graph-building"></a>
</h3>
<p>Next, we can build the optimal graph object. A single function,
called <code>graphBuilder</code>, will be used with subsequent
parameters:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phgObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graphBuilder.html">graphBuilder</a></span><span class="op">(</span></span>
<span>    configFile <span class="op">=</span> <span class="va">config_path</span>,</span>
<span>    methods <span class="op">=</span> <span class="st">"GATK_PIPELINE"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Where:</p>
<ul>
<li>
<code>configFile</code>: the path to our PHG database configuration
file;</li>
<li>
<code>methods</code>: one of the prior methods in the
<code>showMethods()</code> function call, specifically, the
<code>method_name</code> column (<em>see prior output</em>).</li>
</ul>
<p>When the graph has finished building, we can then inspect the
object:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phgObj</span></span></code></pre></div>
<pre><code><span><span class="co">## class: PHGDataSet </span></span>
<span><span class="co">## dim: 20 6 </span></span>
<span><span class="co">## metadata(1): jObj</span></span>
<span><span class="co">## assays(1): hapID</span></span>
<span><span class="co">## rownames(20): R1 R11 ... R10 R20</span></span>
<span><span class="co">## rowData names(1): refRange_id</span></span>
<span><span class="co">## colnames(6): LineA LineA1 ... Ref RefA1</span></span>
<span><span class="co">## colData names(0):</span></span></code></pre>
<p>When built, the object that is generated is of a
<code>PHGDataSet</code> class. In the next section, we will discuss how
to extract specific data from this object.</p>
</div>
</div>
<div class="section level2">
<h2 id="accessing-data">Accessing data<a class="anchor" aria-label="anchor" href="#accessing-data"></a>
</h2>
<div class="section level3">
<h3 id="overview-2">Overview<a class="anchor" aria-label="anchor" href="#overview-2"></a>
</h3>
<p>When called, the <code>PHGDataSet</code> shows the similar output to
a <code>SummarizedExperiment</code> or
<code>RangedSummarizedExperiment</code> class. More information about
this commonly used Bioconductor class can be found <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">here</a>.
This is due to the <code>PHGDataSet</code> class inheriting all methods
and slot information from a <code>SummarizedExperiment</code> class.
Therefore, we can use general accessor methods from either the
<code>SummarizedExperiment</code> or <code>S4Vectors</code> packages.
Some examples are as follows:</p>
<ul>
<li><code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code></li>
<li><code><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges()</a></code></li>
<li><code>metadata()</code></li>
</ul>
<p>A more detailed example of how these methods can be used in a
<code>PHGDataSet</code> are shown below:</p>
<p><img src="img/rphg_data_structure.png" alt="The layout of a PHGDataSet class."><br></p>
<p>The three main data types that can be extracted is:</p>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges()</a></code>; <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>: reference range
data</li>
<li>
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>; <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames()</a></code>: taxa
(i.e. genotype) data</li>
<li>
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay()</a></code>: haplotype or path ID data in matrix form</li>
<li>
<code>metadata()</code>: access to the PHG API Java object (<em>for
advanced use only</em>)</li>
</ul>
</div>
<div class="section level3">
<h3 id="getting-reference-range-data">Getting reference range data<a class="anchor" aria-label="anchor" href="#getting-reference-range-data"></a>
</h3>
<p>Reference range data can be accessed via the <code><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges()</a></code>
method. To use this, simply call this function on the prior PHG object
(e.g. <code>phgObj</code> from out prior examples):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make a `GRanges` object</span></span>
<span><span class="va">rr</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">phgObj</span><span class="op">)</span></span>
<span><span class="va">rr</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 20 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames      ranges strand | refRange_id</span></span>
<span><span class="co">##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt;</span></span>
<span><span class="co">##    R1        1      1-3500      * |          R1</span></span>
<span><span class="co">##   R11        1   3501-7500      * |         R11</span></span>
<span><span class="co">##    R2        1  7501-11000      * |          R2</span></span>
<span><span class="co">##   R12        1 11001-15000      * |         R12</span></span>
<span><span class="co">##    R3        1 15001-18500      * |          R3</span></span>
<span><span class="co">##   ...      ...         ...    ... .         ...</span></span>
<span><span class="co">##   R18        1 56001-60000      * |         R18</span></span>
<span><span class="co">##    R9        1 60001-63500      * |          R9</span></span>
<span><span class="co">##   R19        1 63501-67500      * |         R19</span></span>
<span><span class="co">##   R10        1 67501-71000      * |         R10</span></span>
<span><span class="co">##   R20        1 71001-75000      * |         R20</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>What has been extracted is an object of <code>GRanges</code> class.
This is another commonly used class in the “Bioconductor universe” to
represent genomic intervals. More information about this package can be
found <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" class="external-link">here</a>.
Within this objecct, we get specified reference range information:</p>
<ul>
<li>
<code>seqnames</code>: the chromosome from which this reference
range is found on</li>
<li>
<code>ranges</code>: an <code>IRanges</code> object that contains
reference range coordinates</li>
<li>
<code>strand</code>: which strand for each reference range
(<code>+</code>, <code>-</code>, <code>*</code>)</li>
<li>
<code>refRange_id</code>: given reference range ID within the PHG
database</li>
</ul>
<p>From this object, we can then use additional methods to extract more
specific details such as start/stop coordinates and width of each
reference range:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu">ranges</span><span class="op">(</span><span class="va">rr</span><span class="op">)</span></span></code></pre></div>
<pre><code>IRanges object with 20 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
   R1         1      3500      3500
  R11      3501      7500      4000
   R2      7501     11000      3500
  R12     11001     15000      4000
   R3     15001     18500      3500
  ...       ...       ...       ...
  R18     56001     60000      4000
   R9     60001     63500      3500
  R19     63501     67500      4000
  R10     67501     71000      3500
  R20     71001     75000      4000</code></pre>
<p>Due to the wonderful world of “inheritance”, we can also call this
function directly on <code>SummarizedExperiment</code> objects and
therefore our main <code>PHGDataSet</code> object:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu">ranges</span><span class="op">(</span><span class="va">phgObj</span><span class="op">)</span></span></code></pre></div>
<pre><code>IRanges object with 20 ranges and 0 metadata columns:
          start       end     width
      &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
   R1         1      3500      3500
  R11      3501      7500      4000
   R2      7501     11000      3500
  R12     11001     15000      4000
   R3     15001     18500      3500
  ...       ...       ...       ...
  R18     56001     60000      4000
   R9     60001     63500      3500
  R19     63501     67500      4000
  R10     67501     71000      3500
  R20     71001     75000      4000</code></pre>
</div>
<div class="section level3">
<h3 id="getting-haplotype-data">Getting haplotype data<a class="anchor" aria-label="anchor" href="#getting-haplotype-data"></a>
</h3>
<p>Additionally, haplotype identifiers can also be acessed from this
data object. This is obtained using the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays()</a></code> method,
calling specifically, the haplotype ID matrix (<code>$hapID</code>):</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">phgObj</span><span class="op">)</span><span class="op">$</span><span class="va">hapID</span></span></code></pre></div>
<pre><code><span><span class="co">##     LineA LineA1 LineB LineB1 Ref RefA1</span></span>
<span><span class="co">## R1     41    101    61    121  21    81</span></span>
<span><span class="co">## R11    51    111    71    131  31    91</span></span>
<span><span class="co">## R2     42    102    62    122  22    82</span></span>
<span><span class="co">## R12    52    112    72    132  32    92</span></span>
<span><span class="co">## R3     43    103    63    123  23    83</span></span>
<span><span class="co">## R13    53    113    73    133  33    93</span></span>
<span><span class="co">## R4     44    104    64    124  24    84</span></span>
<span><span class="co">## R14    54    114    74    134  34    94</span></span>
<span><span class="co">## R5     45    105    65    125  25    85</span></span>
<span><span class="co">## R15    55    115    75    135  35    95</span></span>
<span><span class="co">## R6     46    106    66    126  26    86</span></span>
<span><span class="co">## R16    56    116    76    136  36    96</span></span>
<span><span class="co">## R7     47    107    67    127  27    87</span></span>
<span><span class="co">## R17    57    117    77    137  37    97</span></span>
<span><span class="co">## R8     48    108    68    128  28    88</span></span>
<span><span class="co">## R18    58    118    78    138  38    98</span></span>
<span><span class="co">## R9     49    109    69    129  29    89</span></span>
<span><span class="co">## R19    59    119    79    139  39    99</span></span>
<span><span class="co">## R10    50    110    70    130  30    90</span></span>
<span><span class="co">## R20    60    120    80    140  40   100</span></span></code></pre>
<p>The above matrix represents a haplotype ID for each taxa (column
names) within the PHG database at each given reference range (row
names).</p>
</div>
<div class="section level3">
<h3 id="access-java-data">Access Java data<a class="anchor" aria-label="anchor" href="#access-java-data"></a>
</h3>
<p>(<strong>NOTE</strong>: <em>for advanced use only!</em>) To access
Java object data for the PHG API, the <code>metadata()</code> method
from <code>S4Vectors</code> can be used. Doing so will return a Java
reference object that can be used for defining you own potential
functions and methods that do not exist within the <code>rPHG</code>
package:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">phgObj</span><span class="op">)</span><span class="op">$</span><span class="va">jObj</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Java-Object{net.maizegenetics.pangenome.api.HaplotypeGraph@f381794}"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="summary-functions">Summary Functions<a class="anchor" aria-label="anchor" href="#summary-functions"></a>
</h2>
<div class="section level3">
<h3 id="overview-3">Overview<a class="anchor" aria-label="anchor" href="#overview-3"></a>
</h3>
<p>Additionally, <code>rPHG</code> currently has several summary- and
visualization-based functions. These include ways to determine levels of
similarity between reference ranges and taxa.</p>
</div>
<div class="section level3">
<h3 id="determine-numbers-of-haplotypes-per-reference-range">Determine numbers of haplotypes per reference range<a class="anchor" aria-label="anchor" href="#determine-numbers-of-haplotypes-per-reference-range"></a>
</h3>
<p>To see the total number of haplotypes at each given reference range,
the function, <code><a href="../reference/numHaploPerRange.html">numHaploPerRange()</a></code> can be used:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rPHG</span><span class="fu">::</span><span class="fu"><a href="../reference/numHaploPerRange.html">numHaploPerRange</a></span><span class="op">(</span>phgObject <span class="op">=</span> <span class="va">phgObj</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20 rows and 5 columns</span></span>
<span><span class="co">##           names     start       end     width numHaplotypes</span></span>
<span><span class="co">##     &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;     &lt;integer&gt;</span></span>
<span><span class="co">## 1            R1         1      3500      3500             6</span></span>
<span><span class="co">## 2           R11      3501      7500      4000             6</span></span>
<span><span class="co">## 3            R2      7501     11000      3500             6</span></span>
<span><span class="co">## 4           R12     11001     15000      4000             6</span></span>
<span><span class="co">## 5            R3     15001     18500      3500             6</span></span>
<span><span class="co">## ...         ...       ...       ...       ...           ...</span></span>
<span><span class="co">## 16          R18     56001     60000      4000             6</span></span>
<span><span class="co">## 17           R9     60001     63500      3500             6</span></span>
<span><span class="co">## 18          R19     63501     67500      4000             6</span></span>
<span><span class="co">## 19          R10     67501     71000      3500             6</span></span>
<span><span class="co">## 20          R20     71001     75000      4000             6</span></span></code></pre>
<p>The data that is returned will contain each reference range
coordinate, IDs, and the number of haplotypes per range
(<code>numHaplotypes</code>). This data can also be passed to a plotting
function, <code><a href="../reference/plotNumHaplo.html">plotNumHaplo()</a></code> which will display the prior data
as a linear genomic plot. If you are familiar with pipe
(<code>%&gt;%</code>) operators in R via the <code>magrittr</code>
package, this pipeline from <code>PHGDataSet</code> to visualization can
be used as well:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (1) Non-pipe example</span></span>
<span><span class="va">haploPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotNumHaplo.html">plotNumHaplo</a></span><span class="op">(</span><span class="fu"><a href="../reference/numHaploPerRange.html">numHaploPerRange</a></span><span class="op">(</span>phgObject <span class="op">=</span> <span class="va">phgObj</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># (2) Pipe example. Need to load `magrittr` package first!</span></span>
<span><span class="va">haploPlot</span> <span class="op">&lt;-</span> <span class="va">phgObj</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">rPHG</span><span class="fu">::</span><span class="fu"><a href="../reference/numHaploPerRange.html">numHaploPerRange</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">rPHG</span><span class="fu">::</span><span class="fu"><a href="../reference/plotNumHaplo.html">plotNumHaplo</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Return visualization</span></span>
<span><span class="va">haploPlot</span></span></code></pre></div>
<p><img src="img/num_hap_plot.png" alt="Visualize haplotype numbers per reference range."><br></p>
<p>Each reference range is displayed as blue-colored rectangle based on
its physical position within the genome while the y-axis denotes the
total number of unique haplotype IDs per reference range.</p>
</div>
<div class="section level3">
<h3 id="visualize-mutual-information-between-pairs-of-ranges">Visualize mutual information between pairs of ranges<a class="anchor" aria-label="anchor" href="#visualize-mutual-information-between-pairs-of-ranges"></a>
</h3>
<p><code><a href="../reference/plotMutualInfo.html">plotMutualInfo()</a></code> plots the “amount of mutual
information” obtained about one random variable through observing the
other random variable. The higher the numeric value, the more related
taxa for a given reference range.</p>
<p>For this function, a <code>PHGDataSet</code> and reference range IDs
in the form of a vector (e.g. <code>c("R1", "R2", "R3")</code>) are
needed. This function can also be piped using the <code>%&gt;%</code>
operator from <code>magrittr</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">phgObj</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">rPHG</span><span class="fu">::</span><span class="fu"><a href="../reference/plotMutualInfo.html">plotMutualInfo</a></span><span class="op">(</span></span>
<span>      refRanges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">phgObj</span><span class="op">)</span><span class="op">$</span><span class="va">hapID</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="img/corrplot_example.png" alt="Visualize reference range “relatedness”."><br></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p><b>rPHG</b> is an R package developed by <a href="https://www.maizegenetics.net/" class="external-link">The Buckler Lab</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
